sunbeam-python/sunbeam/commands/upgrades/inter_channel.py

```diff
@@ -0,0 +1,336 @@
+# Copyright (c) 2023 Canonical Ltd.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+# implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import logging
+from typing import Callable, Dict, List, Optional, Union
+
+from rich.console import Console
+from rich.status import Status
+
+from sunbeam.clusterd.client import Client
+from sunbeam.commands.juju import JujuStepHelper
+from sunbeam.commands.terraform import TerraformHelper
+from sunbeam.commands.upgrades.base import UpgradeCoordinator, UpgradePlugins
+from sunbeam.commands.upgrades.intra_channel import LatestInChannel
+from sunbeam.jobs.common import (
+    BaseStep,
+    Result,
+    ResultType,
+    read_config,
+    run_plan,
+    update_config,
+)
+from sunbeam.jobs.juju import JujuHelper, run_sync
+from sunbeam.versions import (
+    CHARM_VERSIONS,
+    MACHINE_SERVICES,
+    MISC_SERVICES_K8S,
+    OPENSTACK_SERVICES_K8S,
+    OVN_SERVICES_K8S,
+)
+
+LOG = logging.getLogger(__name__)
+console = Console()
+
+
+class BaseUpgrade(BaseStep, JujuStepHelper):
+    def __init__(self, name, description, jhelper, tfhelper, model):
+        """Create instance of BaseUpgrade class.
+
+        :jhelper: Helper for interacting with pylibjuju
+        :tfhelper: Helper for interaction with Terraform
+        :model: Name of model containing charms.
+        """
+        super().__init__(name, description)
+        self.jhelper = jhelper
+        self.tfhelper = tfhelper
+        self.model = model
+
+    def get_upgrade_strategy(self) -> List[Dict[str, Union[Callable, List]]]:
```

Comment:
This return type is hard to read, the doc in the function is great, but to simplify things, maybe introducing a typed dict might make things simpler ?

```python
from typing import TypedDict, Callable
class UpgradeStrategy(TypedDict):
    upgrade_f: Callable[[list[str], str], None]
    applications: list[str]

def get_upgrade_strategy(self) -> list[UpgradeStrategy]:
```

Reply:
There are other occurrences, like sunbeam-python/sunbeam/plugins/interface/v1/openstack.py L288 and L303